zeus/device/cpu/common.py

```diff
@@ -0,0 +1,150 @@
+"""Error wrappers and classes common to all CPU vendors."""
+
+from __future__ import annotations
+
+import abc
+from typing import Sequence, Optional
+from dataclasses import dataclass
+
+from zeus.device.exception import ZeusBaseCPUError
+
+
+@dataclass
+class CpuDramMeasurement:
+    """Represents a measurement of CPU and DRAM energy consumption.
+
+    Attributes:
+        cpu_mj (int): The CPU energy consumption in millijoules.
+        dram_mj (Optional[int]): The DRAM energy consumption in millijoules. Defaults to None.
+
+    Methods:
+        __sub__(self, other: CpuDramMeasurement) -> CpuDramMeasurement:
+            Subtracts the values of another CpuDramMeasurement from this one.
+
+        __truediv__(self, other: Union[CpuDramMeasurement, float]) -> CpuDramMeasurement:
+            Divides the values of this CpuDramMeasurement by a float.
+    """
+
+    cpu_mj: int
+    dram_mj: Optional[int] = None
+
+    def __sub__(self, other: CpuDramMeasurement) -> CpuDramMeasurement:
+        """Subtracts the values of another CpuDramMeasurement from this one.
+
+        Args:
+            other (CpuDramMeasurement): The other CpuDramMeasurement to subtract.
+
+        Returns:
+            CpuDramMeasurement: A new CpuDramMeasurement with the result of the subtraction.
+        """
+        dram_mj = None
+        if self.dram_mj is not None and other.dram_mj is not None:
+            dram_mj = self.dram_mj - other.dram_mj
+        elif self.dram_mj is not None:
+            dram_mj = self.dram_mj
+        elif other.dram_mj is not None:
+            dram_mj = -other.dram_mj
+        return CpuDramMeasurement(self.cpu_mj - other.cpu_mj, dram_mj)
+
+    def __truediv__(self, other: CpuDramMeasurement | float) -> CpuDramMeasurement:
+        """Divides the values of this CpuDramMeasurement by a float.
+
+        Args:
+            other (float): The float to divide by.
+
+        Returns:
+            CpuDramMeasurement: A new CpuDramMeasurement with the result of the division.
+
+        Raises:
+            ZeroDivisionError: If division by zero is attempted.
+        """
+        if isinstance(other, (int, float)):
+            if other == 0:
+                raise ZeroDivisionError("Division by zero is not allowed")
+            dram_mj = None
+            if self.dram_mj is not None:
+                dram_mj = int(self.dram_mj / other)
+            return CpuDramMeasurement(int(self.cpu_mj / other), dram_mj)
```

Comment:
I don't think it's a good idea to lose precision here by casing to `int`. Rather, I think it's better to type `cpu_mj` and `dram_mj` as `float`.

Assigning an `int` to a `float` field is legal (that is, `a: float = 1` is fine, whereas `b: int = 1.0` is not). See [here](https://peps.python.org/pep-0484/#the-numeric-tower).

Reply:
Would it make since to change the type of `cpu_mj` and `dram_mj` to a float in the `CpuDramMeasurement` data class entirely. The `Measurement` class stores the final measurement as a float so keeping it a float throughout the measurements makes since to me.